draft-ietf-mls-protocol.md

```diff
@@ -1444,110 +1452,169 @@ To authenticate a message from a particular member, signatures are required.
+which generates a sequence of keys and nonces. The group member assigned to that
+leaf uses the j-th key/nonce pair in the sequence to encrypt (using the AEAD)
+the j-th message they send during that epoch. In particular, each key/nonce pair
+MUST NOT be used to encrypt more than one message.
 
-The Application AEAD keying material is generated from the following
-input values:
+More precisely, the initial secret of the ratchet for the group member assigned to the leaf with index i is simply the secret of that leaf.
 
-- The Application Secret value;
-- A purpose value indicating the specific value being generated;
-- The length of the key being generated.
+~~~~
+application_[i]_[0]_secret = astree_node_[i]_secret
+~~~~
 
-Note, that because the identity of the participant using the keys to send data
-is included in the initial Application Secret, all successive updates to the
-Application secret will implicitly inherit this ownership.
+Keys, nonces and secrets of ratchets are derived using HKDF-Expand-Label. The
+context in the call is (a hash of) the current Group state, the index of sender's leaf and the position in that sender's ratchet.
 
-All the traffic keying material is recomputed whenever the underlying
-Application Secret changes.
+~~~~
+struct {
+  opaque gshash<0..2^32-1> = Hash(GroupState_[n]);
+  opaque leaf_index<0..2^16-1>;
+  uint32 ratchet_position;
+} RatchetContext
+~~~~
+
+The identity field is copied from the eponymous field in the BasicCredential
+struct. Further, ratchet_position is initialized to 1 for the first
+HKDF-Expand-Label call of a new ratchet and incremented for each subsequent call
+in that ratchet.
+
+HashRatCont[i, j] denotes the hash of a RatchetContext variable where:
+
+- gshash = Hash(GropuState_[n])
+
+- identity = i
+
+- ratchet_position = j
+
+Using this notation the ratchet for the group member at leaf i is defined to be:
+
+~~~~
+application_[i]_[j]_secret
+      |
+      +--> HKDF-Expand-Label(., "app-nonce", HashRatCont[i,j+1], AEAD.nonceLen)
+      |    = application_[i]_[j+1]_nonce
+      |
+      +--> HKDF-Expand-Label(., "app-key", HashRatCont[i,j+1], AEAD.keyLength)
+      |    = application_[i]_[j+1]_key
+      |
+      V
+HKDF-Expand-Label(., "app-secret", HashRatCont[i,j+1], Hash.length)
+= application_[i]_[j+1]_secret
+~~~~
+
+Here, AEAD.nonceLen and AEAD.keyLength denote the lengths in bytes of the
+nonce and key for the AEAD scheme defined by the ciphersuite.
+
+## Deletion Schedule {#deletion-schedule}
+
+It is important to delete all security sensitive values as soon as they, or
+another value derived from them, is used for encryption or decryption.
+
+More precisely, the values application_[i]\_[j]\_key and
+application\_[i]\_[j]_nonce are said to be "consumed" if they were used either
+to:
+
+* encrypt and send a message or
+* to successfully decrypt and authenticate a received message.
+
+More generally, a secret S is "consumed" if any key, nonce or secret derived
```

Comment:
Agreed on compression encrypt/decrypt into one line.

For the second point I went with "any key, secret or nonce" instead of "any secret" just to be clear that nonce's are included here (as they're normally not thought of as secrets).